velox/common/memory/HashStringAllocator.h

```diff
@@ -516,6 +510,148 @@ class HashStringAllocator : public StreamArena {
   State state_;
 };
 
+/// A ByteInputStream over the data in the range of begin and possible
+/// continuation ranges.
+class HashStringAllocator::InputStream : public ByteInputStream {
+ public:
+  explicit InputStream(const Header* begin)
+      : begin_(const_cast<Header*>(begin)), header_(begin_) {
+    resetRange();
+    current_ = &range_;
+  }
+
+  InputStream(const InputStream& other) {
+    *this = other;
+  }
+
+  InputStream& operator=(const InputStream& other) {
+    begin_ = other.begin_;
+    header_ = other.header_;
+    range_ = other.range_;
+    current_ = &range_;
+    return *this;
+  }
+
+  size_t size() const final {
+    auto* header = begin_;
+    size_t total = 0;
+    for (;;) {
+      total += header->usableSize();
+      if (!header->isContinued()) {
+        break;
+      }
+      header = header->nextContinued();
+    }
+    return total;
+  }
+
+  bool atEnd() const final {
+    return range_.position == range_.size && !header_->isContinued();
+  }
+
+  std::streampos tellp() const final {
+    auto* header = begin_;
+    int64_t pos = 0;
+    while (header != header_) {
+      pos += header->usableSize();
+      header = header->nextContinued();
+    }
+    return pos + range_.position;
+  }
+
+  void seekp(std::streampos pos) final {
+    header_ = begin_;
```

Comment:
updates to header_ and range_ are tightly bound, everywhere you call resetRange(), the line above sets header_ to some new value.

it looks like it'd be easier/safer to make resetRange into resetHeader (or something like that)

  ```
void resetHeader(const Header* header) {
    VELOX_DCHECK_GT(header->usableSize(), 0);
    header_ = header;
    range_.buffer = reinterpret_cast<uint8_t*>(header_->begin());
    range_.size = header_->usableSize();
    range_.position = 0;
  }
```